File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        HashMap<Character,Integer> map = new HashMap<>();
        int left = 0, right = 0, longestSubstringLength = 0;
        while(right < len) {
            char c = s.charAt(right);
            map.put(c, map.getOrDefault(c, 0) + 1);
            if(map.get(c) > 1) {
                while (left < right && map.get(c) > 1) {
                    map.put(s.charAt(left), map.get(s.charAt(left)) - 1);
                    if(map.get(c) == 0) {
                        map.remove(c);
                    }
                    left++;
                }
            }
            longestSubstringLength = Math.max(longestSubstringLength, right - left + 1);
            right++;
        }
        return longestSubstringLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> mp;
        int left = 0, right = 0, lengthLongestSubstring = 0;
        while(right < s.length()){
            mp[s[right]]++;
            
            if(mp[s[right]] > 1){
                while(left < right && mp[s[right]] > 1){
                    mp[s[left]]--;
                    left++;
                }
            }
            lengthLongestSubstring = max(lengthLongestSubstring, right - left + 1);
            right++;
        }
        return lengthLongestSubstring;
    }
};